docs/posts/ibis-analytics/index.qmd

```diff
@@ -1,341 +1,1969 @@
+backend.
+
+The PyPI download data is in Parquet format:
+
+```{.python}
+# imports
+import ibis
+import dagster
+
+from dag import functions as f
+
+
+# assets
+@dagster.asset
+def extract_downloads():
+    """
+    Extract the ingested PyPI downloads data.
+    """
+    downloads = f.clean_data(
+        ibis.read_parquet("data/ingest/pypi/ibis-framework/*.parquet")
+    )
+    return downloads
+```
+
+While the GitHub (omitted for brevity) and Zulip data is in JSON format:
+
+```{.python}
+# imports
+import ibis
+import dagster
+
+from dag import functions as f
+
+
+# assets
+@dagster.asset
+def extract_zulip_members():
+    """
+    Extract the ingested Zulip members data.
+    """
+    members = f.clean_data(ibis.read_json("data/ingest/zulip/members.json"))
+    return members
+
+
+@dagster.asset
+def extract_zulip_messages():
+    """
+    Extract the ingested Zulip messages data.
+    """
+    messages = f.clean_data(ibis.read_json("data/ingest/zulip/messages.json"))
+    return messages
+```
+
+As shown above, Dagster assets are configured to use the DuckDB table manager so
+the output of these functions are written to separate DuckDB database files for
+use downstream in the DAG.
+
+### Transform
+
+With the data extracted, we can now transform it into its desired shape for
+downstream analytics. The most interesting transformation code is the PyPI data,
+shown here:
+
+```{.python}
+# imports
+import ibis
+import dagster
+
+from dag import functions as f
+
+
+# assets
+@dagster.asset
+def transform_downloads(extract_downloads):
+    """
+    Transform the PyPI downloads data.
+    """
+    downloads = f.clean_data(
+        extract_downloads.drop("project").unpack("file").unpack("details")
+    )
+    downloads = downloads.mutate(
+        major_minor_patch=f.clean_version(downloads["version"], patch=True),
+        major_minor=f.clean_version(downloads["version"], patch=False).cast("float"),
+    )
+    downloads = downloads.rename(
+        {
+            "version_raw": "version",
+            "version": "major_minor_patch",
+        }
+    )
+    downloads = (
+        downloads.group_by(
+            [
+                ibis._.timestamp.truncate("D").name("timestamp"),
+                ibis._.country_code,
+                ibis._.version,
+                ibis._.python,
+                ibis._.system["name"].name("system"),
+            ]
+        )
+        .agg(
+            ibis._.count().name("downloads"),
+        )
+        .order_by(ibis._.timestamp.desc())
+        .mutate(
+            ibis._.downloads.sum()
+            .over(
+                rows=(0, None),
+                group_by=["country_code", "version", "python", "system"],
+                order_by=ibis._.timestamp.desc(),
+            )
+            .name("total_downloads")
+        )
+        .order_by(ibis._.timestamp.desc())
+    )
+    downloads = downloads.mutate(ibis._["python"].fillna("").name("python_full"))
+    downloads = downloads.mutate(
```

Comment:
All this reassignment to the same variable suggests that perhaps you should chain methods together.

You don't lose any debuggability, since you can successively comment parts of the chain in or out.

Reply:
🤷 it's how the code is currently written and works